ocaml/typing/jkind.ml

```diff
@@ -1752,12 +1574,505 @@ module Type = struct
+            subjkind_format "is not" l2,
+            None )
+        | Some p ->
+          ( l1,
+            l2,
+            dprintf "an unknown layout",
+            subjkind_format "might not be" l2,
+            Some p ))
+      | { violation = No_intersection (l1, l2); missing_cmi } ->
+        assert (Option.is_none missing_cmi);
+        ( l1,
+          l2,
+          dprintf "layout %a" format l1,
+          dprintf "does not overlap with %a" format l2,
+          None )
+      | { violation = No_union (l1, l2); missing_cmi } ->
+        assert (Option.is_none missing_cmi);
+        ( l1,
+          l2,
+          dprintf "layout %a" format l1,
+          dprintf "cannot be summed with %a" format l2,
+          None )
+    in
+    if Type.display_histories
+    then
+      let connective =
+        match t.violation, to_const l2 with
+        | Not_a_subjkind _, Some _ -> dprintf "be a sublayout of %a" format l2
+        | No_intersection _, Some _ -> dprintf "overlap with %a" format l2
+        | No_union _, Some _ -> dprintf "sum with %a" format l2
+        | _, None -> dprintf "be representable"
+      in
+      fprintf ppf "@[<v>%a@;%a@]"
+        (format_history
+           ~intro:(dprintf "The layout of %a is %a" pp_former former format l1))
+        l1
+        (format_history
+           ~intro:
+             (dprintf "But the layout of %a must %t" pp_former former connective))
+        l2
+    else
+      fprintf ppf "@[<hov 2>%s%a has %t,@ which %t.@]" preamble pp_former former
+        fmt_l1 fmt_l2;
+    Type.report_missing_cmi ppf missing_cmi_option
+
+  let report_general preamble pp_former former ppf t =
+    try report_general_type_jkind preamble pp_former former ppf t
+    with Match_failure _ -> fprintf ppf "((arrow violation))"
+
+  let pp_t ppf x = fprintf ppf "%t" x
+
+  let report_with_offender ~offender = report_general "" pp_t offender
+
+  let report_with_offender_sort ~offender =
+    report_general "A representable layout was expected, but " pp_t offender
+
+  let report_with_name ~name = report_general "" pp_print_string name
+end
+
+(******************************)
+(* relations *)
+
+let rec equate_or_equal ~allow_mutation (t : t) (t' : t) =
+  match t, t' with
+  | Type ty, Type ty' -> Type.equate_or_equal ~allow_mutation ty ty'
+  | ( Arrow { args = args1; result = result1 },
+      Arrow { args = args2; result = result2 } ) ->
+    let arg_eqs = List.map2 (equate_or_equal ~allow_mutation) args1 args2 in
+    equate_or_equal ~allow_mutation result1 result2
+    && List.for_all (fun x -> x) arg_eqs
+  | _ -> false
+
+let equate = equate_or_equal ~allow_mutation:true
+
+(* This mirrors the choice for Jkind.Type, following the note:
+   (layouts v2.8) allow_mutation is to be set to false *)
+let equal = equate_or_equal ~allow_mutation:true
+
+(* Generalises union and intersection at Arrows, as they are mutually recursive when
+   defining Arrow kind intersection *)
```

Comment:
I think this set of functions would be easier to follow if written in a more direct style. Something like:
```ocaml
let rec intersection_or_error ~reason k1 k2 =
  match k1, k2 with
  | Type ty1, Type ty2 -> begin
      match Type.intersection_or_error ~reason ty1 ty2 with
      | Ok ty -> Ok (Type ty)
      | Error _ as err -> err
    end
  | Arrow(args1, res1), Arrow(args1, res1) -> begin
      match union_or_error_list ~reason args1 args2 with
      | Ok args -> begin
          match intersection_or_error ~reason res1 res2 with
           | Ok res -> Ok (Arrow { arg; result })
           | Error _ as err -> err
         end
       | Error _ as err -> err
     end
   | (Type _ | Arrow _), _ -> Error (Violation.of_ (No_intersection (k1, k2)))
   
and intersection_or_error_list ~reason arr1 ks1 arr2 ks2 =
  match ts1, ts2 with
  | [], [] -> Ok []
  | k1 :: rest1, k2 :: rest2 -> begin
      match intersection_or_error ~reason k1 k2 with
      | Ok k -> begin
          match intersection_or_error_list ~reason rest1 rest2 with
          | Ok rest -> Ok (k :: rest)
          | Error _ as err -> err
        end
      | Error _ as err -> err
    end
  | [], _ :: _ -> Error (Violation.of_ (No_intersection (arr1, arr2)))
  | _ :: _, [] -> Error (Violation.of_ (No_intersection (arr1, arr2)))
...
```
This code is very simple and clear and requires much less thought to see it is correct. It be even nicer using binding operators, but I'm not sure anyone has really used those in the compiler yet.

Reply:
I attempted a refactor in a more direct style with some `Result.bind` with let operators (I wasn't happy with the nested pattern matches). I do find myself unsatisfied with the code repetition (2 times 2), so I did this generically in the simpler case of combining lists:

```ocaml
let rec combine_list_or_error ~violation combine ts1 ts2 =
  let ( let* ) = Result.bind in
  match ts1, ts2 with
  | [], [] -> Ok []
  | t1 :: ts1, t2 :: ts2 ->
    let* t = combine t1 t2 in
    let* ts = combine_list_or_error ~violation combine ts1 ts2 in
    Ok (t :: ts)
  | [], _ :: _ | _ :: _, [] -> Error violation
```

Perhaps this could be further simplified by factoring out the `map2`, but perhaps this is fine.
I kept `intersection_or_error` and `union_or_error` separate, though.

Reply:
The let operator is definitely nicer. I'm a little nervous of being the first to add it to the compiler code base (I think), but it's so much nicer let's just do it anyway.

Personally I'd prefer to just write `intersection_list_or_error` and `union_list_or_error` rather than using a higher-order combinator. First-order code is just less effort to read and review, and there isn't much benefit from sharing the shared pieces across the two definitions here. But I'm fine with keeping this version.

Reply:
There seem to be sporadic uses of `let*` in `ocaml/parsing/builtin_attributes.ml` and `ocaml/utils/diffing.ml`; and of `let+` in `ocaml/lambda/tmc.ml` and `ocaml/lambda/value_rec_compiler.ml`. I'm not sure all of these are actively used, but I do notice CI passes so hopefully that takes the minimum compiler version into account.